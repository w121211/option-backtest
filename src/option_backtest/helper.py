from datetime import datetime


def generate_output_dir(
    config: dict,
    output_dir: str,
    include_params=["stock_id", "s0", "mu", "sigma_stock", "start_date", "end_date"],
    exclude_params: set[str] = set(),
) -> str:
    output_dir_parts = []

    for param in include_params:
        if param not in exclude_params:
            value = config.get(param)
            if isinstance(value, datetime):
                value = value.strftime("%Y%m%d")
            elif isinstance(value, float):
                value = str(int(value * 100))
            output_dir_parts.append(f"{param}_{value}")

    # output_dir = (
    #     output_dir
    #     # + datetime.now().strftime("%Y%m%d_%H%M%S_")
    #     + "_".join(output_dir_parts)
    # )
    output_dir = f"{output_dir}/{'_'.join(output_dir_parts)}"

    return output_dir
