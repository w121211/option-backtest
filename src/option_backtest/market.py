import copy
import logging
from dataclasses import dataclass
from typing import Literal, Protocol

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from scipy.stats import norm


logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)


@dataclass
class Stock:
    id: str
    df: pd.DataFrame  # Columns = ['price']. Indexed by 'date'
    position: int = 0
    # entry_date: datetime | None = None
    # exit_date: datetime | None = None

    def __repr__(self):
        return f"Stock(id='{self.id}', position={self.position})"


@dataclass
class Option:
    id: str
    option_type: Literal["call", "put"]
    strike: float
    expiry_date: pd.Timestamp
    df: pd.DataFrame  # Columns ['price', 'delta', 'vega', ...], index by 'date'

    position: int = 0
    # entry_date: datetime | None = None
    # exit_date: datetime | None = None

    def __repr__(self):
        return f"Option(id='{self.id}', position={self.position})"


@dataclass
class Account:
    cash: float
    stock: Stock  # Underlying stock
    options: dict[str, Option]  # key: option id, value: option


@dataclass
class Trade:
    asset_type: Literal["stock", "option"]
    asset_id: str
    amount: int
    date: pd.Timestamp

@dataclass
class MarketState:
    stock: Stock  # Underlying stock
    listed_options: dict[str, Option]
    account: Account
    current_date: pd.Timestamp
    current_step: int


class OnTradeFn(Protocol):
    def __call__(self, state: MarketState) -> list[Trade]: ...


def execute_trades(
    state: MarketState,
    trades: list[Trade],
) -> Account:
    listed_options = state.listed_options
    account = state.account
    current_date = state.current_date

    for trade in trades:
        if trade.asset_type == "stock":
            stock = account.stock
            stock_price = stock.df.at[current_date, "price"]
            stock.position += trade.amount
            account.cash -= stock_price * trade.amount

            logger.info(
                f"Executed trade [{current_date.strftime("%Y-%m-%d")}] {trade.asset_id} @ {stock_price:.2f} * {trade.amount}, Stock position: {stock.position}, Cash: {account.cash:.2f}"
            )

        elif trade.asset_type == "option":
            if trade.asset_id not in account.options:
                opt = copy.copy(listed_options[trade.asset_id])
                opt.position = 0
                account.options[opt.id] = opt

            opt = account.options[trade.asset_id]
            opt_price = opt.df.at[current_date, "price"]
            opt.position += trade.amount
            account.cash -= opt_price * trade.amount * 100

            logger.info(
                f"Executed trade[{current_date.strftime("%Y-%m-%d")}] {trade.asset_id} @ {opt_price:.2f} * {trade.amount}, Option position: {opt.position}, Cash: {account.cash:.2f}"
            )

    return account


def after_market(
    state: MarketState
) -> tuple[dict[str, Option], Account]:
    listed_options = state.listed_options
    current_date = state.current_date
    account = state.account

    # Remove expired options from available options
    listed_options = {
        opt_id: opt
        for opt_id, opt in listed_options.items()
        if opt.expiry_date > current_date
    }

    # Check account options
    for opt_id, opt in list(account.options.items()):
        if opt.expiry_date <= current_date:
            current_stock_price = account.stock.df.at[current_date, "price"]

            # Option has expired
            if opt.option_type == "call":
                if current_stock_price > opt.strike and opt.position != 0:
                    # Exercise call option
                    account.stock.position += opt.position * 100
                    account.cash -= opt.position * opt.strike * 100
                    logger.info(
                        f"Exercised call option {opt_id}. Bought {opt.position * 100} shares at strike price {opt.strike}. Current stock price={current_stock_price}"
                    )

            elif opt.option_type == "put":
                if current_stock_price < opt.strike and opt.position != 0:
                    # Exercise put option
                    account.stock.position -= opt.position * 100
                    account.cash += opt.position * opt.strike * 100
                    logger.info(
                        f"Exercised put option {opt_id}. Sold {opt.position * 100} shares at strike price {opt.strike}."
                    )

            # Remove expired option from account
            del account.options[opt_id]
            # logger.info(f"Removed expired option {opt_id} from account.")

    return listed_options, account


def calculate_portfolio_value(account: Account, current_date: pd.Timestamp) -> float:
    portfolio_value = account.cash

    # Calculate stock value
    stock_price = account.stock.df.at[current_date, "price"]
    stock_value = account.stock.position * stock_price
    portfolio_value += stock_value

    # Calculate option values
    for opt in account.options.values():
        if opt.expiry_date > current_date:
            opt_price = opt.df.at[current_date, "price"]
            opt_value = opt.position * opt_price * 100
            portfolio_value += opt_value

    return portfolio_value


def episode(
    init_cash: float,
    listed_stock: Stock,
    listed_options: dict[str, Option],
    on_trade_fns: list[OnTradeFn],  # A list of trading functions
):
    account = Account(cash=init_cash, stock=listed_stock, options={})

    log_trades = []
    log_portfolio_values = []

    for i, date in enumerate(account.stock.df.index):
        state = MarketState(
            stock=listed_stock,
            listed_options=listed_options,
            account=account,
            current_date=date,
            current_step=i,
        )

        trades = []
        for on_trade in on_trade_fns:
            _trades = on_trade(state)
            trades.extend(_trades)

        account = execute_trades(state, trades)
        listed_options, account = after_market(state)
        portfolio_value = calculate_portfolio_value(account, date)

        log_trades.extend(trades)
        log_portfolio_values.append(portfolio_value)

    portfolio_df = pd.DataFrame(
        {"value": log_portfolio_values}, index=account.stock.df.index
    )

    return log_trades, portfolio_df
