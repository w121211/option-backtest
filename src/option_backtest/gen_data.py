from datetime import datetime, timedelta
from typing import Dict, List, Literal, Tuple, TypedDict, Union

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from scipy.stats import norm

from .market import Option, Stock


# -------- Generate Market Data --------


def daily_returns(prices: np.ndarray) -> np.ndarray:
    daily_returns = np.diff(prices) / prices[:-1]
    return daily_returns


def prices_sigma(prices: np.ndarray) -> float:
    returns = daily_returns(prices)
    sigma = np.std(returns) * np.sqrt(252)  # Annualized volatility
    return sigma


def black_scholes_call(S: float, K: float, T: float, r: float, sigma: float) -> float:
    d1 = (np.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
    d2 = d1 - sigma * np.sqrt(T)
    call_price = S * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    return call_price


def call_prices(
    prices: np.ndarray,
    strikes: List[float],
    days_to_expire: List[int],
    sigma: float,
    r: float = 0.05,
) -> pd.DataFrame:
    n_days = len(prices)
    results = []
    for strike in strikes:
        for days in days_to_expire:
            call_prices = []
            for i in range(n_days):
                S = prices[i]
                T = (days - i) / 252  # Update time to expiration
                if T >= 0:
                    call_price = black_scholes_call(S, strike, T, r, sigma)
                    call_prices.append(call_price)
                else:
                    call_prices.append(np.nan)  # Or any other appropriate value
            call_prices = np.array(call_prices)
            daily_pct_changes = np.diff(call_prices) / call_prices[:-1]
            daily_pct_changes = np.insert(
                daily_pct_changes, 0, np.nan
            )  # Insert 'NA' for day 0
            results.extend(
                zip(
                    np.repeat(days, n_days),
                    np.repeat(strike, n_days),
                    range(n_days),
                    call_prices,
                    daily_pct_changes,
                )
            )
    df = pd.DataFrame(
        results,
        columns=["Expire", "Strike", "Day", "Call Price", "Call Pct Change"],
    )
    return df


def gbm_prices(
    s0: float, mu: float, sigma: float, dt: float, n_steps: int
) -> np.ndarray:
    z = np.random.normal(size=n_steps)
    prices = s0 * np.exp(
        np.cumsum((mu - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * z)
    )
    return prices


def generate_and_save_stock_data(
    s0: float,
    mu: float,
    sigma: float,
    dt: float,
    start_date: datetime,
    end_date: datetime,
    output_path: str,
):
    n_steps = (end_date - start_date).days + 1
    prices = gbm_prices(s0, mu, sigma, dt, n_steps)

    # Generate dates
    dates = [start_date + timedelta(days=int(i)) for i in range(n_steps)]

    # Save stock prices to CSV
    stock_data = pd.DataFrame({"date": dates, "price": prices})
    stock_data.to_csv(output_path, index=False)


def generate_and_save_option_data(
    prices, start_date, r, sigma, strikes, expiry_dates, output_path
):
    n_steps = len(prices)
    df = call_prices(
        prices,
        strikes,
        [int((expiry_date - start_date).days) for expiry_date in expiry_dates],
        sigma,
        r,
    )

    # Generate dates
    dates = [start_date + timedelta(days=int(i)) for i in range(n_steps)]

    # Save option data to separate CSVs
    for strike in strikes:
        for expiry_date in expiry_dates:
            expiry_date_str = expiry_date.strftime("%y%m%d")
            strike_str = str(int(strike * 1000)).zfill(8)
            option_name = f"XYZ{expiry_date_str}C{strike_str}"
            days = (expiry_date - start_date).days
            option_data = df[(df["Strike"] == strike) & (df["Expire"] == days)]
            option_data = option_data[["Day", "Call Price", "Call Pct Change"]]
            option_data["Date"] = [dates[int(i)] for i in option_data["Day"]]
            option_data = option_data[["Date", "Call Price", "Call Pct Change"]]
            option_data.to_csv(f"{output_path}/{option_name}.csv", index=False)


def plot_asset_price(file_path):
    data = pd.read_csv(file_path)
    data["Date"] = pd.to_datetime(data["Date"])

    fig = go.Figure()
    fig.add_trace(
        go.Scatter(x=data["Date"], y=data["Price"], mode="lines", name="Price")
    )
    fig.update_layout(
        title="Asset Price Over Time", xaxis_title="Date", yaxis_title="Price"
    )
    fig.show()


# -------- Black Scholes Model --------


def black_scholes(S, K, T, r, sigma, option_type: Literal["call", "put"]):
    """
    Calculate the Black-Scholes option price for a single set of input parameters.

    Parameters:
    - S (float): Current stock price.
    - K (float): Strike price of the option.
    - T (float): Time to expiration (in years).
    - r (float): Risk-free interest rate (annualized).
    - sigma (float): Volatility of the underlying stock (annualized).
    - option_type (str): Type of the option - "call" or "put".

    Returns:
    - float: Calculated option price.
    """
    d1 = (np.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
    d2 = d1 - sigma * np.sqrt(T)

    if option_type.lower() == "call":
        price = S * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    elif option_type.lower() == "put":
        price = K * np.exp(-r * T) * norm.cdf(-d2) - S * norm.cdf(-d1)
    else:
        raise ValueError("Invalid option_type. Must be 'call' or 'put'.")

    return price


def black_scholes_vectorized_old(
    S: Union[np.ndarray, pd.Series],
    K: float,
    T: Union[np.ndarray, pd.Series],
    r: float,
    sigma: float,
    option_type: str,
) -> Union[np.ndarray, pd.Series]:
    """
    Vectorized Black-Scholes formula, modified to handle cases where T <= 0
    by setting the option price to zero.

    Parameters:
    - S: Stock price (np.ndarray or pd.Series)
    - K: Strike price (float)
    - T: Time to maturity in years (np.ndarray or pd.Series)
    - r: Risk-free interest rate (float)
    - sigma: Volatility of the underlying asset (float)
    - option_type: 'call' or 'put' (str)

    Returns:
    - Option price (np.ndarray or pd.Series)
    """

    # Initialize option_price as zeros
    if isinstance(T, np.ndarray):
        option_price = np.zeros_like(T, dtype=np.float64)
    elif isinstance(T, pd.Series):
        option_price = pd.Series(np.zeros(len(T)), dtype=np.float64, index=T.index)

    # Ensure T is non-negative, else set option price to 0
    mask_T = T > 0
    S_filtered = S[mask_T]  # Filter S as well to match the filtered T's dimensions
    T_filtered = T[mask_T]

    if np.any(mask_T):  # Proceed only if there are any valid T values
        d1 = (np.log(S_filtered / K) + (r + 0.5 * sigma**2) * T_filtered) / (
            sigma * np.sqrt(T_filtered)
        )
        d2 = d1 - sigma * np.sqrt(T_filtered)

        if option_type == "call":
            option_price[mask_T] = S_filtered * norm.cdf(d1) - K * np.exp(
                -r * T_filtered
            ) * norm.cdf(d2)
        elif option_type == "put":
            option_price[mask_T] = K * np.exp(-r * T_filtered) * norm.cdf(
                -d2
            ) - S_filtered * norm.cdf(-d1)
        else:
            raise ValueError("Invalid option_type. Must be 'call' or 'put'.")
    # For T <= 0, option_price remains zero as initialized

    return option_price


def black_scholes_vectorized(
    S: Union[np.ndarray, pd.Series],
    K: float,
    T: Union[np.ndarray, pd.Series],
    r: float,
    sigma: float,
    option_type: str,
) -> Tuple[Union[np.ndarray, pd.Series], Union[np.ndarray, pd.Series]]:
    """
    Vectorized Black-Scholes formula, modified to handle cases where T <= 0
    by setting the option price and delta to zero.

    Parameters:
    - S: Stock price (np.ndarray or pd.Series)
    - K: Strike price (float)
    - T: Time to maturity in years (np.ndarray or pd.Series)
    - r: Risk-free interest rate (float)
    - sigma: Volatility of the underlying asset (float)
    - option_type: 'call' or 'put' (str)

    Returns:
    - Option price (np.ndarray or pd.Series)
    - Option delta (np.ndarray or pd.Series)
    """

    # Initialize option_price and option_delta as zeros
    if isinstance(T, np.ndarray):
        option_price = np.zeros_like(T, dtype=np.float64)
        option_delta = np.zeros_like(T, dtype=np.float64)
    elif isinstance(T, pd.Series):
        option_price = pd.Series(np.zeros(len(T)), dtype=np.float64, index=T.index)
        option_delta = pd.Series(np.zeros(len(T)), dtype=np.float64, index=T.index)

    # Ensure T is non-negative, else set option price and delta to 0
    mask_T = T > 0
    S_filtered = S[mask_T]  # Filter S as well to match the filtered T's dimensions
    T_filtered = T[mask_T]

    if np.any(mask_T):  # Proceed only if there are any valid T values
        d1 = (np.log(S_filtered / K) + (r + 0.5 * sigma**2) * T_filtered) / (
            sigma * np.sqrt(T_filtered)
        )
        d2 = d1 - sigma * np.sqrt(T_filtered)

        if option_type == "call":
            option_price[mask_T] = S_filtered * norm.cdf(d1) - K * np.exp(
                -r * T_filtered
            ) * norm.cdf(d2)
            option_delta[mask_T] = norm.cdf(d1)
        elif option_type == "put":
            option_price[mask_T] = K * np.exp(-r * T_filtered) * norm.cdf(
                -d2
            ) - S_filtered * norm.cdf(-d1)
            option_delta[mask_T] = norm.cdf(d1) - 1
        else:
            raise ValueError("Invalid option_type. Must be 'call' or 'put'.")
    # For T <= 0, option_price and option_delta remain zero as initialized

    return option_price, option_delta


def calculate_option_price(
    stock_price_df: pd.DataFrame,
    expiry_date: pd.Timestamp,
    strike: float,
    option_type: str,
    r: float,
    sigma: float,
    trading_days_per_year: int = 252,
) -> pd.DataFrame:
    """
    Calculate option prices for a given set of stock prices using the Black-Scholes model.

    Parameters:
    - stock_prices: DataFrame containing stock prices with columns ['price'] and indexed by 'date'
    - expiry_date: Expiration date of the option (datetime object)
    - strike: Strike price of the option
    - option_type: Type of the option ('call' or 'put')
    - r: Risk-free interest rate (annualized)
    - sigma: Volatility of the underlying asset (annualized)
    - trading_days_per_year: Number of trading days in a year (default: 252)

    Returns:
    - DataFrame containing option data with columns ['price', 'delta'] and indexed by 'date'
    """
    # Calculate time to maturity for each stock price date
    time_to_maturity = (
        expiry_date - stock_price_df.index
    ).to_series().dt.days / trading_days_per_year

    # Calculate option prices using the Black-Scholes model
    option_price, option_delta = black_scholes_vectorized(
        stock_price_df["price"].values,
        strike,
        time_to_maturity.values,
        r,
        sigma,
        option_type,
    )

    # Create a new DataFrame with date and option price columns
    option_df = pd.DataFrame(
        {
            # "date": stock_price_df["date"],
            "price": option_price,
            "delta": option_delta,
        },
        index=stock_price_df.index,
    )
    # option_price_df = option_price_df.set_index("date")
    return option_df


# ------------------------------------------------------------
# Option
# ------------------------------------------------------------


def generate_options(
    stock: Stock,
    strikes: np.ndarray,
    expiry_dates: list[pd.Timestamp],
    r: float,
    sigma: float,
) -> dict[str, Option]:
    """
    Generate a dictionary of call and put options for the given stock, with the specified strike prices and expiry dates.

    The option ID is constructed following the convention described in:
    https://polygon.io/blog/how-to-read-a-stock-options-ticker

    Parameters:
    - stock: The underlying Stock object for which the options are generated.
    - strikes: A NumPy array of strike prices for the options.
    - expiry_dates: A list of expiry dates for the options, represented as pd.Timestamp objects.
    - r: The risk-free interest rate used for option pricing.
    - sigma: The volatility of the underlying stock used for option pricing.

    Returns:
    - A dictionary of Option objects, where the keys are the option IDs and the values are the corresponding Option objects.
    """
    options: dict[str, Option] = {}
    for expiry_date in expiry_dates:
        for opt_type in ["call", "put"]:
            for strike in strikes:
                expiry_date_str = expiry_date.strftime("%y%m%d")
                strike_price_str = f"{int(abs(strike) * 1000):08d}"
                opt_id = f"{stock.id}{expiry_date_str}{'C' if opt_type == 'call' else 'P'}{strike_price_str}"
                opt_df = calculate_option_price(
                    stock.df, expiry_date, strike, opt_type, r, sigma
                )
                opt = Option(
                    id=opt_id,
                    option_type=opt_type,
                    strike=strike,
                    expiry_date=expiry_date,
                    df=opt_df,
                )
                options[opt.id] = opt
    return options
