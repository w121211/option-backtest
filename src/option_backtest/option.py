from datetime import datetime, timedelta
from typing import Callable, Literal, TypedDict
from typing_extensions import NotRequired
import pandas as pd

from .market import Option


FindOptionsBy = TypedDict(
    "FindOptionsBy",
    {
        # Filter options by the "call" or "put".
        "option_type": Literal["call", "put"],
        # Filter options by the "days to expiration" greter than x days.
        "dte_gt": NotRequired[int],
        # Filter options by the "days to expiration" less than x days.
        "dte_lt": NotRequired[int],
        # Filter options by the delta is greater than x. Use this to filter ITM options by delta > 0.5.
        "delta_gt": NotRequired[float],
        # Filter options by the "days to expiration" within a range (min_days, max_days).
        "dte_range": NotRequired[tuple[int, int]],
        # Filter options by the delta is less than x. Use this to filter OTM options by delta < 0.5.
        "delta_lt": NotRequired[float],
        # Sort by delta near x
        "sort_by_delta_near": NotRequired[float],
    },
)


def find_options(
    listed_options: dict[str, Option],
    current_stock_price: float,
    current_date: pd.Timestamp,
    attrs: FindOptionsBy,
) -> list[Option]:
    filtered_options = list(listed_options.values())

    if attrs.get("option_type"):
        filtered_options = [
            op for op in filtered_options if op.option_type == attrs["option_type"]
        ]

    if attrs.get("dte_gt"):
        dte_gt = attrs["dte_gt"]  # type: ignore
        filtered_options = [
            op
            for op in filtered_options
            if (op.expiry_date - current_date).days > dte_gt
        ]

    if attrs.get("dte_lt"):
        dte_lt = attrs["dte_lt"]  # type: ignore
        filtered_options = [
            op
            for op in filtered_options
            if (op.expiry_date - current_date).days < dte_lt
        ]

    if attrs.get("dte_range"):
        min_days, max_days = attrs["dte_range"]  # type: ignore
        filtered_options = [
            op
            for op in filtered_options
            if min_days <= (op.expiry_date - current_date).days < max_days
        ]

    if attrs.get("delta_gt"):
        delta_gt = attrs["delta_gt"]  # type: ignore
        filtered_options = [
            op for op in filtered_options if op.df.at[current_date, "delta"] > delta_gt
        ]
        filtered_options = sorted(
            filtered_options, key=lambda op: op.df.at[current_date, "delta"]
        )

    if attrs.get("delta_lt"):
        delta_lt = attrs["delta_lt"]  # type: ignore
        filtered_options = [
            op for op in filtered_options if op.df.at[current_date, "delta"] < delta_lt
        ]
        filtered_options = sorted(
            filtered_options,
            key=lambda op: op.df.at[current_date, "delta"],
            reverse=True,
        )

    # Sort options by delta near the specified value
    if attrs.get("sort_by_delta_near"):
        delta_near = attrs["sort_by_delta_near"]  # type: ignore

        def sort_key(op):
            try:
                return abs(op.df.at[current_date, "delta"] - delta_near)
            except KeyError:
                return float("inf")

        filtered_options = sorted(filtered_options, key=sort_key)

    return filtered_options


# FindOptionsCondition = tuple[
#     Literal["delta", "dte", "option_type"], Callable[[float | str], bool]
# ]


# def find_options(
#     listed_options: dict[str, Option],
#     current_stock_price: float,
#     current_date: pd.Timestamp,
#     conditions: list[FindOptionsCondition],
#     sort_by_delta_near: float | None = None,
# ) -> list[Option]:
#     def apply_conditions(op: Option) -> bool:
#         for cond_type, cond_func in conditions:
#             if cond_type == "delta":
#                 if not cond_func(op.df.at[current_date, "delta"]):
#                     return False
#             elif cond_type == "dte":
#                 dte = (op.expiry_date - current_date).days
#                 if not cond_func(dte):
#                     return False
#             elif cond_type == "option_type":
#                 if not cond_func(op.option_type):
#                     return False
#         return True

#     filtered_options = [op for op in listed_options.values() if apply_conditions(op)]

#     if sort_by_delta_near is not None:
#         filtered_options = sorted(
#             filtered_options,
#             key=lambda op: abs(op.df.at[current_date, "delta"] - sort_by_delta_near),
#         )

#     return filtered_options


def should_rollover(option: Option, current_date: datetime) -> bool:
    """Determine if the held option should be rolled over based on the current date."""
    return (option.expiry_date - current_date).days <= 10


def find_next_option(
    available_options: list[Option], current_date: datetime, current_market_price: float
) -> Option | None:
    """Find the next option to rollover to based on the criteria:
    1. Has at least 30 days before expiry.
    2. Is nearest to the current day.
    3. Is ATM (At-The-Money).
    """
    # Filter options that have at least 30 days before expiry
    valid_options = [
        option
        for option in available_options
        if (option.expiry_date - current_date).days >= 30
    ]
    if not valid_options:
        return None

    # Finding the option that is nearest to being At-The-Money (ATM)
    atm_option = min(
        valid_options,
        key=lambda x: (
            abs(x.strike - current_market_price),
            (x.expiry_date - current_date).days,
        ),
    )
    return atm_option


def rollover_options(
    available_options: list[Option], prices_df: pd.DataFrame
) -> list[Option]:
    prices_df["date"] = pd.to_datetime(prices_df["date"])

    rolled_over_options = []
    current_option = None

    for i, row in prices_df.iterrows():
        current_date = row["date"]
        current_market_price = row["price"]

        if current_option is None or should_rollover(current_option, current_date):
            next_option = find_next_option(
                available_options, current_date, current_market_price
            )
            if next_option and (
                current_option is None or next_option.id != current_option.id
            ):
                # if current_option:
                #     current_option.exit_date = current_date
                # next_option.entry_date = current_date

                rolled_over_options.append(next_option)
                print(
                    f"{current_date.strftime('%y-%m-%d')} {current_market_price} => Rolled over to option {next_option.strike}@{next_option.expiry_date.strftime('%y%m%d')}"
                )
                current_option = next_option  # Update the held option
            elif not next_option:
                print("No suitable option found to rollover to.")
                break  # Exit the loop if no suitable next option is found

    return rolled_over_options


def rollover_short_atm_calls(
    available_options: list[Option], prices_df: pd.DataFrame
) -> list[Option]:
    """
    Implement a short ATM (At-The-Money) call option rollover strategy.

    The strategy sells the nearest ATM call option with at least 30 days to expiry
    and rolls over to the next ATM call option when there are 10 days or fewer left to expiry.

    Parameters:
    - available_options: List of available Option objects to choose from.
    - prices_df: DataFrame with columns ['date', 'price'] representing the underlying asset prices.

    Returns:
    - List of Option objects representing the rolled-over short call options.
    """
    prices_df["date"] = pd.to_datetime(prices_df["date"])

    rolled_over_options = []
    current_option = None

    for i, row in prices_df.iterrows():
        current_date = row["date"]
        current_market_price = row["price"]

        if (
            current_option is None
            or (current_option.expiry_date - current_date).days <= 10
        ):
            # Find the next ATM call option to sell
            valid_options = [
                option
                for option in available_options
                if (option.expiry_date - current_date).days >= 30
                and option.option_type == "call"
            ]
            if valid_options:
                next_option = min(
                    valid_options,
                    key=lambda x: (
                        abs(x.strike - current_market_price),
                        (x.expiry_date - current_date).days,
                    ),
                )
                if current_option is None or next_option.id != current_option.id:
                    # if current_option:
                    #     current_option.exit_date = current_date
                    # current_option.position = 0  # Close the current short position
                    # next_option.entry_date = current_date
                    next_option.position = -1  # Open a new short position

                    rolled_over_options.append(next_option)
                    print(
                        f"{current_date.strftime('%y-%m-%d')} {current_market_price} => Rolled over to selling option {next_option.strike}@{next_option.expiry_date.strftime('%y%m%d')}"
                    )
                    current_option = next_option  # Update the sold option
            else:
                print("No suitable option found to rollover to.")
                break  # Exit the loop if no suitable next option is found

    return rolled_over_options
