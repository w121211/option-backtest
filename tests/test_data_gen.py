from datetime import datetime
import numpy as np
import pandas as pd
import pytest
from scipy.stats import norm

from src.option_backtest.market import Stock
from src.option_backtest.gen_data import (
    black_scholes_vectorized,
    generate_options,
)


# -------- Test Black Scholes ------


def test_black_scholes_vectorized_call():
    S = np.array([100, 100, 100])
    K = 100
    T = np.array([1, 0.5, 0])
    r = 0.05
    sigma = 0.2
    option_type = "call"

    expected_price = np.array([10.4506, 6.8887, 0])
    expected_delta = np.array([0.6368, 0.5977, 0])

    price, delta = black_scholes_vectorized(S, K, T, r, sigma, option_type)

    np.testing.assert_almost_equal(price, expected_price, decimal=4)
    np.testing.assert_almost_equal(delta, expected_delta, decimal=4)


def test_black_scholes_vectorized_put():
    S = np.array([100, 100, 100])
    K = 100
    T = np.array([1, 0.5, 0])
    r = 0.05
    sigma = 0.2
    option_type = "put"

    expected_price = np.array([5.5735, 4.4197, 0])
    expected_delta = np.array([-0.3632, -0.4023, 0])

    price, delta = black_scholes_vectorized(S, K, T, r, sigma, option_type)

    np.testing.assert_almost_equal(price, expected_price, decimal=4)
    np.testing.assert_almost_equal(delta, expected_delta, decimal=4)


def test_black_scholes_vectorized_pandas():
    S = pd.Series([100, 100, 100])
    K = 100
    T = pd.Series([1, 0.5, 0])
    r = 0.05
    sigma = 0.2
    option_type = "call"

    expected_price = pd.Series([10.4506, 6.8887, 0])
    expected_delta = pd.Series([0.6368, 0.5977, 0])

    price, delta = black_scholes_vectorized(S, K, T, r, sigma, option_type)

    pd.testing.assert_series_equal(price, expected_price, atol=1e-4)
    pd.testing.assert_series_equal(delta, expected_delta, atol=1e-4)


def test_black_scholes_vectorized_invalid_option_type():
    S = np.array([100])
    K = 100
    T = np.array([1])
    r = 0.05
    sigma = 0.2
    option_type = "invalid"

    with pytest.raises(
        ValueError, match="Invalid option_type. Must be 'call' or 'put'."
    ):
        black_scholes_vectorized(S, K, T, r, sigma, option_type)


def test_black_scholes_vectorized_with_nonpositive_time_to_maturity():
    S = np.array([100, 100, 100])
    K = 100
    T = np.array([1, 0, -0.5])  # Mix of positive, zero, and negative time to maturity
    r = 0.05
    sigma = 0.2
    option_type = "call"

    expected_price = np.array([10.4506, 0, 0])
    expected_delta = np.array([0.6368, 0, 0])

    price, delta = black_scholes_vectorized(S, K, T, r, sigma, option_type)

    np.testing.assert_almost_equal(price, expected_price, decimal=4)
    np.testing.assert_almost_equal(delta, expected_delta, decimal=4)


@pytest.mark.parametrize(
    "T_values, expected_results",
    [
        (
            # T values
            np.array([-1, 0, 0.5, 1, 2]),
            # Expected option prices
            np.array([0.0, 0.0, 6.88872858, 10.45058357, 16.12677972]),
        ),
        (
            # T values reorder
            np.array([0.5, -1, 1, 0, 2]),
            # Expected option prices
            np.array([6.88872858, 0.0, 10.45058357, 0.0, 16.12677972]),
        ),
    ],
)
def test_black_scholes_time_ranges(T_values, expected_results):
    S = np.array([100, 100, 100, 100, 100])  # Constant stock price for simplicity
    K = 100  # Strike price
    r = 0.05  # Risk-free interest rate
    sigma = 0.2  # Volatility
    option_type = "call"  # Testing with call option

    # Call the vectorized function
    price, delta = black_scholes_vectorized(S, K, T_values, r, sigma, option_type)

    # Compare the expected and calculated prices
    np.testing.assert_allclose(
        price,
        expected_results,
        rtol=1e-5,
        atol=1e-5,
        err_msg="Vectorized results should match expected true values",
    )


# -------- Test Generate Options ------

# def test_generate_options(snapshot):
#     # Define test input parameters
#     strikes = np.array([50, 60, 70])
#     expiry_dates = [
#         datetime(2023, 6, 30),
#         datetime(2023, 9, 30),
#     ]
#     underlying_stock_ticker = "XYZ"

#     # Call the function to generate options
#     options = generate_options(strikes, expiry_dates, underlying_stock_ticker)

#     assert options == snapshot


def test_generate_options(snapshot):
    # Create a sample stock object
    stock_df = pd.DataFrame(
        {"price": [100, 110, 120]},
        index=pd.date_range(start="2023-01-01", periods=3, freq="d"),
    )
    stock = Stock(id="XYZ", df=stock_df)

    # Define the input parameters
    strikes = np.array([90, 100, 110])
    expiry_dates = [pd.Timestamp("2023-06-30"), pd.Timestamp("2023-12-31")]
    r = 0.05
    sigma = 0.2

    # Call the generate_options function
    options = generate_options(stock, strikes, expiry_dates, r, sigma)
    opt_0_df = list(options.values())[0].df

    assert options == snapshot
    assert opt_0_df.to_string() == snapshot
