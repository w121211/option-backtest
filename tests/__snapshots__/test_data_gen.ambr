# serializer version: 1
# name: test_generate_options
  dict({
    'XYZ230630C00090000': Option(id='XYZ230630C00090000', position=None),
    'XYZ230630C00100000': Option(id='XYZ230630C00100000', position=None),
    'XYZ230630C00110000': Option(id='XYZ230630C00110000', position=None),
    'XYZ230630P00090000': Option(id='XYZ230630P00090000', position=None),
    'XYZ230630P00100000': Option(id='XYZ230630P00100000', position=None),
    'XYZ230630P00110000': Option(id='XYZ230630P00110000', position=None),
    'XYZ231231C00090000': Option(id='XYZ231231C00090000', position=None),
    'XYZ231231C00100000': Option(id='XYZ231231C00100000', position=None),
    'XYZ231231C00110000': Option(id='XYZ231231C00110000', position=None),
    'XYZ231231P00090000': Option(id='XYZ231231P00090000', position=None),
    'XYZ231231P00100000': Option(id='XYZ231231P00100000', position=None),
    'XYZ231231P00110000': Option(id='XYZ231231P00110000', position=None),
  })
# ---
# name: test_generate_options.1
  '''
                  price     delta
  2023-01-01  14.934464  0.820985
  2023-01-02  23.741388  0.931291
  2023-01-03  33.301941  0.977553
  '''
# ---
